Make MeshBVH.closestPointToPoint, closestPointToGeometry return more info

yomboprime opened the pull request:
fixes #318, #321 

I've made the changes described in the mentioned issues: Target parameters can be Vector3 (as before), or objects (either ```{ }``` or the object from a previous execution). The objects (if not null) are filled with face indices, normal, uv, etc.

For ```closestPointToGeometry``` I've made it return always the info for target1 (```this.geometry```), but target2 (related to otherGeometry) is updated if it is an object. Mixing Vector3 and info object for target1 and target2 is not suported.

Also I've removed the ```distanceToPoint``` and ```distanceToGeometry``` functions of ```MeshBVH```.

The code has become large, let me know if there is something you would remove/modify.

Edit: Forgot to mention I've run the tests and they work fine. But the benchmark seems not to work, it complains about ```performance.now()``` not being defined.

yomboprime commented:
I was creating new vectors instead of reusing the target objects ones. I fixed it in the last commit, bloating even more the code :-/

yomboprime commented:
I think at this point it could be better to make two versions of each function, for performance reasons and readability. What do you think?

gkjohnson commented on src/core/MeshBVH.js:
```diff
@@ -943,13 +1084,74 @@ export class MeshBVH {
+
+			}
+
+			if ( ! target.point ) target.point = new Vector3();
+			target.point.copy( temp1 );
+			target.distance = closestDistance;
+			if ( ! target.face ) target.face = { };
+			target.face.a = a;
+			target.face.b = b;
+			target.face.c = c;
+			target.face.materialIndex = 0;
+			if ( ! target.face.normal ) target.face.normal = new Vector3();
+			Triangle.getNormal( tempV1, tempV2, tempV3, target.face.normal );
+			if ( ! target.uv ) target.uv = new Vector2();
+			target.uv.copy( uv );
+
+			return target;
```
A lot of this looks like it could be reused between the two cases in `closestPointToGeometry` and this function -- should we make a utility function for constructing this based on a point, geometry, and triangle index? It could go in the [utils/TriangleUtilities.js file](https://github.com/gkjohnson/three-mesh-bvh/blob/master/src/utils/TriangleUtilities.js). The function could modify an existing object if it's passed in or return a new one if there's no object provided.

gkjohnson reviewed (commented):
This is great! Thank you so much.

> The code has become large, let me know if there is something you would remove/modify.

I've left a couple comments on the `closestPointToGeometry` but I think the big thing is that constructing the "hit" information object could be done in a utility function to cut down on the logic within each function. And the backwards compatibility / deprecation warnings can happen in a wrapped function, though that doesn't seem to be taking up a lot of space in the functions.

> Edit: Forgot to mention I've run the tests and they work fine. But the benchmark seems not to work, it complains about performance.now() not being defined.

Yes `performance.now` was added in node v16 so I switched over to that from Date.now() for more fidelity.

yomboprime commented:
I've made a mix of your suggestions and another approach.

I've extracted the hit info collection code in a function as you said, but I don't call it from ```closestPointToPoint``` nor ```closestPointToGeometry```. Instead, I've changed the result and target objects of those two Functions to just contain the point, the distance and the triangle index. Then, the caller can use the utility triangle function afterwards. This saves time when the caller doesn't need hit info, and there is no penalty when hit info is needed.

The utility function is ```getTriangleHitPointInfo()```

I've modified the tests. I can also edit the docs if you think this approach is right.

gkjohnson commented on src/core/MeshBVH.js:
```diff
@@ -943,13 +964,27 @@ export class MeshBVH {
+		const closestDistance = Math.sqrt( closestDistanceSq );
 
-	}
+		if ( target ) {
+
+			if ( ! target.point ) target.point = temp1.clone();
+			else target.point.copy( temp1 );
+			target.distance = closestDistance,
+			target.faceIndex = closestDistanceTriIndex;
+			return target;
+
+		} else {
```
I think it's okay to default `target` to an empty object in the constructor which will simplify final condition block further, as well. Then we won't have to handle the "else" condition because it would be guaranteed that the target object is defined.

We can do the same for `closestPointToGeometry` for at least `target1`.

gkjohnson commented on src/core/MeshBVH.js:
```diff
@@ -787,19 +789,21 @@ export class MeshBVH {
 										if ( dist < closestDistance ) {
 
-											if ( target1 ) {
+											if ( tempTargetDest1 ) {
```
`tempTargetDest1` will always be defined, now.

gkjohnson reviewed (commented):
> I've made a mix of your suggestions and another approach.
> ...
> The utility function is `getTriangleHitPointInfo()`

Interesting! I kinda like the idea. I think in practice calling that function won't be adding a lot of overhead compared to the bounds traversal but lets go with this. Let's just make sure we point to the function from `closestPointToGeometry` and `closestPointToPoint`. If it winds up posing a problem we can always change it easily later.

>  I can also edit the docs if you think this approach is right.

Updating the docs would be great. And lets also make sure the function is exposed  through `index.js` so it's public.

That aside I added a couple more comments. Thanks for the changes so far! I think these should be the last adjustments.

yomboprime commented on src/core/MeshBVH.js:
```diff
@@ -875,13 +881,31 @@ export class MeshBVH {
+		if ( target2 ) {
 
-	}
+			if ( ! target2.point ) target2.point = tempTargetDest2.clone();
+			else target2.point.copy( tempTargetDest2 );
+			target2.distance = closestDistance,
+			target2.faceIndex = closestDistanceOtherTriIndex;
```
Yes, I think ```target2.point``` should be converted to the other geometry frame. But the distance should remain the same, no need to recompute it?
Sorry for not spotting the frame issue.

gkjohnson replied:
> But the distance should remain the same, no need to recompute it?

If the geometries are scaled differently to one another then the distance represented in each local frame would be different. Ie if "otherGeometry" were scaled to `0.1` while the current geometry were scaled to `10` then there's a scale difference of 100 between them meaning the distance would be different a factor of 100.

Not all transforms will have uniforms scales, though. It would probably be easiest to transform both points into the second geometry in order to recompute the distance.

> Sorry for not spotting the frame issue.

Ha I've been working on this project for awhile now and there are still things I'm surprised I didn't see previously!

yomboprime commented:
> Interesting! I kinda like the idea. I think in practice calling that function won't be adding a lot of overhead compared to the bounds traversal but lets go with this. Let's just make sure we point to the function from `closestPointToGeometry` and `closestPointToPoint`. If it winds up posing a problem we can always change it easily later.
> 

Sorry, I don't understand here. What do you mean by 'pointing'? A comment in the code or in the docs?

> Updating the docs would be great. And lets also make sure the function is exposed through `index.js` so it's public.
> 

Sorry, I didn't know exposing in ```index.js``` was neccesary. I will do it.

gkjohnson commented:
> Sorry, I don't understand here. What do you mean by 'pointing'? A comment in the code or in the docs?

I just mean a link in the docs from each functions description so users know they can use this new function to compute the remaining hit information.

yomboprime commented:
Done!

The only issue I've found is: I had to set targets to ```undefined``` so they get the default objects. With ```null``` they stay null.

yomboprime commented:
Should be done now. I also updated the docs.

gkjohnson commented on benchmark/run-benchmark.js:
```diff
@@ -18,8 +18,8 @@ box.min.set( 1, 1, 1 );
 
-const target1 = new THREE.Vector3();
-const target2 = new THREE.Vector3();
+const target1 = undefined;
+const target2 = undefined;
```
Lets set these to empty objects {} so the objects will be reused and to reduce the impact of garbage collection on the benchmark.

gkjohnson commented on README.md:
```diff
@@ -499,6 +488,41 @@ getBoundingBox( target : Box3 ) : Box3
+
+- `point`: The same point
+- `a`, `b`, `c`: Triangle indices
+- `materialIndex`: Face material index or 0 if not available.
```
From the function implementation it looks like the materialIndex is always set to 0, right? Technically the material index should be based on the `geometry.groups` list, but a better approach will require changes to the "shapecast" function so the current group id is passed down to the callbacks, which I'll make an issue for.

gkjohnson replied:
#325

gkjohnson reviewed (commented):
One more thing! (and a last lint issue to fix) After these we can merge it.

Also sorry for the delay on the workflow results. It's requiring that I approve to run the tests on the PR since it's your first contribution.

yomboprime commented:
Sorry, I forgot to push the last commit. It should be all done now.

gkjohnson commented:
This is great, thank you again!
